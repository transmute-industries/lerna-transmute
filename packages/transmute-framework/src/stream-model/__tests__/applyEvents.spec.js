const { EventStore } = require("../../index");
const transmuteConfig = require("../../transmute-config");
const abi = require("../../../build/contracts/EventStore.json");
const Web3 = require("web3");
const TransmuteAdapterIPFS = require("transmute-adapter-ipfs");
const provider = new Web3.providers.HttpProvider(
  transmuteConfig.web3Config.providerUrl
);
const web3 = new Web3(provider);
const adapter = new TransmuteAdapterIPFS(transmuteConfig.ipfsConfig);
let eventStore = new EventStore({
  web3,
  abi,
  adapter
});

const StreamModel = require("../index");

const events = require("../../__mock__/events.json");

describe("applyEvents", () => {
  let accounts;
  let streamModel;

  const filter = event => {
    return event;
  };
  const reducer = (state, event) => {
    return {
      ...state,
      events: [event.value, ...(state.events || [])],
      eventsProcessed: (state.eventsProcessed || 0) + 1
    };
  };

  beforeAll(async () => {
    eventStore = new EventStore({
      web3,
      abi,
      adapter
    });
    await eventStore.init();
    accounts = await eventStore.getWeb3Accounts();
  });

  // new event store per test
  beforeEach(async () => {
    eventStore = await eventStore.clone(accounts[0]);
    streamModel = new StreamModel(eventStore, filter, reducer);
  });

  it("should be idempotent", async () => {
    streamModel.applyEvents(events);
    expect(streamModel.state.model).toMatchSnapshot();
    streamModel.applyEvents(events);
    expect(streamModel.state.model).toMatchSnapshot();
  });

  it("should support custom event filters", async () => {
    const custom_filter = event => {
      return event.key.type === "patient" && event.key.id === "0";
    };
    const streamModel = new StreamModel(eventStore, custom_filter, reducer);
    streamModel.applyEvents(events);
    expect(streamModel.state.model).toMatchSnapshot();
  });
});
