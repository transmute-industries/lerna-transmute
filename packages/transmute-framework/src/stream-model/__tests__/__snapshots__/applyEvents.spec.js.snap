// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyEvents should be idempotent 1`] = `Object {}`;

exports[`applyEvents should be idempotent 2`] = `Object {}`;

exports[`applyEvents should support custom event filters 1`] = `
Object {
  "events": Array [
    Object {
      "index": "1",
      "key": Object {
        "id": "0",
        "type": "patient",
      },
      "value": Object {
        "type": "USER_REGISTERED",
        "username": "bob@example.com",
      },
    },
    Object {
      "index": "3",
      "key": Object {
        "id": "0",
        "type": "patient",
      },
      "value": Object {
        "guid": "0x0...",
        "type": "DEVICE_LINKED",
      },
    },
    Object {
      "index": "4",
      "key": Object {
        "id": "0",
        "type": "patient",
      },
      "value": Object {
        "type": "USER_UPDATED",
        "username": "bob2@example.com",
      },
    },
  ],
}
`;
